src/js/reducers/conversation-reducer.js

```diff
@@ -36,22 +34,75 @@ const addMessage = (messages, message) => {
     return sortMessages([...messages, message]);
 };
 
+
+const matchMessage = (message, queryProps) => Object.keys(queryProps).every((key) => message[key] === queryProps[key]);
+
+const replaceMessage = (messages, query, newMessage) => {
+    const existingMessage = messages.find((message) => matchMessage(message, query));
+    if (!existingMessage) {
+        return messages;
+    }
+
+    const index = messages.indexOf(existingMessage);
+    return [...messages.slice(0, index), newMessage, ...messages.slice(index + 1)];
+};
+
+const isEqual = (a, b) => {
+    if (a._id && b._id && a._id === b._id) {
+        return true;
+    }
+
+    if (!a._id || !b._id) {
+        if (a.role === b.role) {
+            if (a.text && b.text && a.text === b.text) {
+                return true;
+            }
+
+            if (a.mediaType === b.mediaType && a.mediaUrl === b.mediaUrl) {
+                return true;
+            }
+        }
+    }
+
+    return false;
+};
+
+const mergeMessages = (a, b) => {
+    // concat will make a union out of both arrays
+    return removeDuplicates(a.concat(b));
+};
+
+const removeDuplicates = (array) => {
```

Comment:
replaceMessage already checks if the message is there or not before it proceeds. Do we need this ~O(n^2) removeDuplicates check?